### ISSUE TITLE
Unexpected Timezone offset with ical4j 4.0.0-rc1

### ISSUE THREAD

**Issue by hprange** [ical4j/ical4j#662]
**Describe the bug**
I've been experimenting with the latest release candidate of ical4j and I've noticed some peculiar behavior when formatting `ZonedDateTime` instances with the `America/Sao_Paulo` timezone.

**To Reproduce**
Here's a snippet of code that generates a `VEvent` for a one-hour meeting using the aforementioned timezone:

```java
var startDate = ZonedDateTime.of(2023, 11, 15, 8, 30, 0, 0, ZoneId.of("America/Sao_Paulo"));
var endDate = startDate.plusHours(1);

System.out.print(new VEvent(startDate, endDate, "Metting"));
```

**Expected behavior**
The output I'm getting is:

```
startDate ==> 2023-11-15T08:30-03:00[America/Sao_Paulo]
endDate ==> 2023-11-15T09:30-03:00[America/Sao_Paulo]

BEGIN:VEVENT
DTSTAMP:20231118T152115Z
DTSTART;TZID=America/Sao_Paulo:20231115T093000
DTEND;TZID=America/Sao_Paulo:20231115T103000
SUMMARY:Metting
END:VEVENT
```

As you can see, the offset for this datetime with `America/Sao_Paulo` is -03:00. However, the `VEvent` seems to be using a -02:00 offset for the start and end dates. It appears as though daylight savings time is being applied, but [Brazil abolished daylight savings in 2019](https://www.timeanddate.com/time/change/brazil).

**Environment:**
 - OS: macOS 14.1.1
 - Java: 15.0.2
 - iCal4j: 4.0.0-rc1

**Additional context**
I've tried enabling the `net.fortuna.ical4j.timezone.update.enabled` option to no avail. I also verified the [timezone definition at tzurl.org](https://www.tzurl.org/zoneinfo/America/Sao_Paulo) and it seems to be correct.

I'd appreciate any insights or suggestions you might have. Thanks!

**Comment by benfortuna** [comment]
I have confirmed the historical timezone implementation was wrong. I have now reimplemented `ZoneRulesBuilder` and added some tests to confirm it now works.

Fix available in 4.0.0-rc2

**Comment by haavar** [comment]
I tested with 4.0.0-rc2 and it's working for me now.
